proposals/NNNN-nonescapable-stdlib-primitives.md

```diff
@@ -0,0 +1,871 @@
+
+Important convenience APIs such as `Result.init(catching:)` or `Result.map` will need to require escapability until we introduce a way to formally specify lifetime dependencies. This is unfortunate, but it still enables intrepid Swift developers to experiment with defining interfaces that take (or even return!) `Result` values.
+
+However, we are already able to generalize a small handful of methods: `get` and the two error-mapping utilities, `mapError` and `flatMapError`.
+
+```swift
+func sample<E: Error>(_ res: Result<Span<Int>, E>) -> Int {
+  guard let span = try? res.get() else { return 42 }
+  return 3 * span.count + 9
+}
+```
+
+Like unwrapping an `Optional`, calling `get()` on a nonescapable `Result` returns a value whose lifetime requirements exactly match that of the original `Result` instance -- the act of unwrapping a result cannot allow its content to escape its intended context.
+
+### Retrieving the memory layout of nonescapable types
+
+This proposal generalizes `enum MemoryLayout` to support retrieving information about the layout of nonescapable types:
+
+```swift
+print(MemoryLayout<Span<Int>>.size) // ⟹ 16
+print(MemoryLayout<Span<Int>>.stride) // ⟹ 16
+print(MemoryLayout<Span<Int>>.alignment) // ⟹ 8
+```
+
+(Of course, the values returned will vary depending on the target architecture.)
+
+The information returned is going to be of somewhat limited use until we generalize unsafe pointer types to support nonescapable pointees, which this proposal does not include -- but there is no reason to delay this work until then.
+
+(To usefully allow pointers to nonescapable types, we'll need to assign precise lifetime semantics to their `pointee` (and pointer dereferencing in general), and we'll most likely also need a way to allow developers to unsafely override the resulting default lifetime semantics. This requires explicit lifetime annotations, and as such, that work is postponed to a future proposal.)
+
+### Lifetime management
+
+We once again generalize the `withExtendedLifetime` family of functions, this time to support calling them on nonescapable values.
+
+```swift
+let span = someArray.storage
+withExtendedLifetime(span) { span in
+  // `someArray` is being actively borrowed while this closure is running
+}
+// At this point, `someArray` may be ready to be mutated
+```
+
+We've now run proposals to generalize `withExtendedLifetime` for (1) typed throws, (2) noncopyable inputs and results, and (3) nonescapable inputs. It is getting unwieldy to keep having to tweak these APIs, especially since in actual practice, `withExtendedLifetime` is most often called with an empty closure, to serve as a sort of fence protecting against early destruction.
+
+To acknowledge this widespread pattern, and to avoid having to keep indefinitely generalizing these marginal functions, we propose to introduce a simple `extendLifetime` function that avoids 
```

Comment:
```suggestion
To acknowledge this widespread pattern, and to avoid having to keep indefinitely generalizing these marginal functions, we propose to introduce a simple `extendLifetime` function that avoids this closure argument.
```

Reply:
Good catch! The Detailed Design had some duplicate text explaining the same points; I fixed this by unifying the two sections, removing the rationale from the latter.